recovery state: fix concurrent access to file list

brwe opened the pull request:
A shard recovery response might serialize a shard state at the same time that it is
modified by the recovery process. The test
RelocationTests.testMoveShardsWhileRelocation
failed because of this with a ConcurrentModificationException.

bleskes commented on src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java:
```diff
@@ -872,7 +872,7 @@ public void readFrom(StreamInput in) throws IOException {
         public void writeTo(StreamOutput out) throws IOException {
             super.writeTo(out);
-            final File[] files = fileDetails.values().toArray(new File[0]);
+            final File[] files = fileDetails().toArray(new File[0]);
```
GOOD CATCH. I think the right solution is to synchronize the method , just as all other methods are. Can you also do it to the readFrom?

bleskes commented on src/test/java/org/elasticsearch/indices/recovery/RecoveryStateTest.java:
```diff
@@ -481,4 +484,46 @@ Start createObj() {
     }
+
+    @Test
+    public void testConcurrentAccessToIndexFileDetails() throws InterruptedException {
```
I wonder why the testIndex that has a concurrent reader didn't fail?

brwe replied:
It was not removing entries from the map or adding any while the streamer wrote. The exception came from the call to AbstractCollection.toArray in writeTo

http://build-us-00.elastic.co/job/es_core_master_strong/3228/testReport/junit/org.elasticsearch.recovery/RelocationTests/testMoveShardsWhileRelocation/

bleskes commented:
good catch! left one comment about the solution, and a question

brwe commented:
Thanks for the review! Addressed all comments. I also changed the test now to use the Streamer class that was already there instead of a custom thread.

bleskes commented:
LGTM. in my opinion this should go to 1.5.1

s1monw commented:
> LGTM. in my opinion this should go to 1.5.1

+1

brwe commented:
This is already fixed on 1.5 and 1.x by 3a9a33ec4c7f4daffe26b4b1dc39f156108fce39 the commit to master (8c695355807be445d0322ecffb28377c1117c7a9) was just missing the synchronized. I will just push the test to 1.5 and 1.x.
